File: ts/test/session.test.ts
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { MantyxClient } from "../src/index.js";
import { MockServer } from "./helpers/mock-server.js";

let server: MockServer;
let client: MantyxClient;

beforeEach(async () => {
  server = new MockServer();
  await server.start();
  client = new MantyxClient({
    apiKey: "test-key",
    workspaceSlug: "demo",
    baseUrl: server.baseUrl(),
  });
});

afterEach(async () => {
  await server.stop();
});

describe("AgentSession", () => {
  it("create + send round-trips a multi-turn conversation", async () => {
    const session = await client.createSession({ systemPrompt: "you are a chat bot" });
    server.scriptForNextSessionRun = {
      events: [{ type: "result", subtype: "success", text: "first reply" }],
    };
    const out1 = await session.send("first");
    expect(out1.text).toBe("first reply");

    server.scriptForNextSessionRun = {
      events: [{ type: "result", subtype: "success", text: "second reply" }],
    };
    const out2 = await session.send("second");
    expect(out2.text).toBe("second reply");

    const history = await session.history();
    expect(history.map((h) => h.content)).toEqual([
      "first",
      "first reply",
      "second",
      "second reply",
    ]);

    await session.end();
  });

  it("forwards createSession metadata and per-message metadata overrides", async () => {
    await client.createSession({
      systemPrompt: "you are a chat bot",
      metadata: { customer: "acme", env: "prod" },
    });
    expect(server.lastSessionCreateBody?.metadata).toEqual({
      customer: "acme",
      env: "prod",
    });

    const session = await client.createSession({ systemPrompt: "x" });
    server.scriptForNextSessionRun = {
      events: [{ type: "result", subtype: "success", text: "ok" }],
    };
    await session.send("hello", { metadata: { trace_id: "trace_123" } });
    expect(server.lastSessionMessageBody?.metadata).toEqual({ trace_id: "trace_123" });
  });

  it("lists sessions and filters by metadata", async () => {
    await client.createSession({
      systemPrompt: "x",
      metadata: { customer: "acme", env: "prod" },
    });
    await client.createSession({
      systemPrompt: "x",
      metadata: { customer: "globex", env: "prod" },
    });

    const all = await client.listSessions();
    expect(all.total).toBe(2);

    const firstPage = await client.listSessions({ limit: 1 });
    expect(firstPage.sessions).toHaveLength(1);
    expect(firstPage.nextCursor).toBeTruthy();
    const secondPage = await client.listSessions({
      limit: 1,
      cursor: firstPage.nextCursor!,
    });
    expect(secondPage.sessions).toHaveLength(1);
    expect(secondPage.sessions[0]?.sessionId).not.toBe(
      firstPage.sessions[0]?.sessionId,
    );
    expect(secondPage.nextCursor).toBeNull();

    const filtered = await client.listSessions({
      metadata: { customer: "acme" },
    });
    expect(filtered.total).toBe(1);
    expect(filtered.sessions[0]?.metadata).toEqual({
      customer: "acme",
      env: "prod",
    });
    expect(filtered.sessions[0]?.status).toBe("active");
    expect(typeof filtered.sessions[0]?.creationDate).toBe("string");

    const none = await client.listSessions({
      metadata: { customer: "acme", env: "staging" },
    });
    expect(none.total).toBe(0);
  });

  it("replays a session as realtime-style event frames", async () => {
    const session = await client.createSession({ systemPrompt: "x" });
    server.scriptForNextSessionRun = {
      events: [{ type: "result", subtype: "success", text: "two" }],
    };
    await session.send("one");
    server.scriptForNextSessionRun = {
      events: [{ type: "result", subtype: "success", text: "four" }],
    };
    await session.send("three");

    const full = await client.getSessionEvents(session.id);
    expect(full).toEqual([
      { seq: 1, type: "user_message", text: "one" },
      { seq: 2, type: "assistant_message", text: "two" },
      { seq: 3, type: "user_message", text: "three" },
      { seq: 4, type: "assistant_message", text: "four" },
    ]);

    const lastTwo = await client.getSessionEvents(session.id, {
      lastMessages: 2,
    });
    expect(lastTwo).toEqual([
      { seq: 3, type: "user_message", text: "three" },
      { seq: 4, type: "assistant_message", text: "four" },
    ]);

    const newestPage = await session.eventsPage({ lastMessages: 2 });
    expect(newestPage.events).toEqual(lastTwo);
    expect(newestPage.nextBeforeSeq).toBe(3);
    expect(newestPage.truncated).toBe(true);

    const olderPage = await session.eventsPage({
      beforeSeq: newestPage.nextBeforeSeq!,
    });
    expect(olderPage.events).toEqual([
      { seq: 1, type: "user_message", text: "one" },
      { seq: 2, type: "assistant_message", text: "two" },
    ]);
    expect(olderPage.nextBeforeSeq).toBeNull();
    expect(olderPage.truncated).toBe(false);
  });
});

File: go/session_test.go
package mantyx

import (
	"context"
	"encoding/json"
	"testing"
)

func TestSession_MultiTurn(t *testing.T) {
	server := newMockServer()
	defer server.close()
	client := NewClient(Options{
		APIKey:        "k",
		WorkspaceSlug: "demo",
		BaseURL:       server.baseURL(),
	})

	session, err := client.CreateSession(context.Background(), SessionSpec{SystemPrompt: "be helpful"})
	if err != nil {
		t.Fatalf("CreateSession: %v", err)
	}
	out1, err := session.Send(context.Background(), "first")
	if err != nil {
		t.Fatalf("Send #1: %v", err)
	}
	if out1.Text != "echo:first" {
		t.Fatalf("unexpected reply 1: %q", out1.Text)
	}
	out2, err := session.Send(context.Background(), "second")
	if err != nil {
		t.Fatalf("Send #2: %v", err)
	}
	if out2.Text != "echo:second" {
		t.Fatalf("unexpected reply 2: %q", out2.Text)
	}
	hist, err := session.History(context.Background())
	if err != nil {
		t.Fatalf("History: %v", err)
	}
	if len(hist) != 4 {
		t.Fatalf("expected 4 history entries, got %d (%v)", len(hist), hist)
	}
	if err := session.End(context.Background()); err != nil {
		t.Fatalf("End: %v", err)
	}
}

func TestSession_MetadataForwarded(t *testing.T) {
	server := newMockServer()
	defer server.close()
	client := NewClient(Options{
		APIKey:        "k",
		WorkspaceSlug: "demo",
		BaseURL:       server.baseURL(),
	})

	session, err := client.CreateSession(context.Background(), SessionSpec{
		SystemPrompt: "be helpful",
		Metadata:     map[string]string{"customer": "acme", "env": "prod"},
	})
	if err != nil {
		t.Fatalf("CreateSession: %v", err)
	}
	var createBody map[string]any
	if err := json.Unmarshal(server.lastSessionCreateBody, &createBody); err != nil {
		t.Fatalf("parse create body: %v", err)
	}
	meta, _ := createBody["metadata"].(map[string]any)
	if meta["customer"] != "acme" || meta["env"] != "prod" {
		t.Fatalf("create metadata not forwarded: %#v", createBody["metadata"])
	}

	if _, err := session.Send(context.Background(), "hi", WithMetadata(map[string]string{"trace_id": "trace_1"})); err != nil {
		t.Fatalf("Send: %v", err)
	}
	var msgBody map[string]any
	if err := json.Unmarshal(server.lastSessionMessageBody, &msgBody); err != nil {
		t.Fatalf("parse message body: %v", err)
	}
	msgMeta, _ := msgBody["metadata"].(map[string]any)
	if msgMeta["trace_id"] != "trace_1" {
		t.Fatalf("send metadata not forwarded: %#v", msgBody["metadata"])
	}
}

func TestListSessions_FiltersByMetadata(t *testing.T) {
	server := newMockServer()
	defer server.close()
	client := NewClient(Options{
		APIKey:        "k",
		WorkspaceSlug: "demo",
		BaseURL:       server.baseURL(),
	})
	ctx := context.Background()

	if _, err := client.CreateSession(ctx, SessionSpec{
		SystemPrompt: "x",
		Metadata:     map[string]string{"customer": "acme", "env": "prod"},
	}); err != nil {
		t.Fatalf("CreateSession #1: %v", err)
	}
	if _, err := client.CreateSession(ctx, SessionSpec{
		SystemPrompt: "x",
		Metadata:     map[string]string{"customer": "globex", "env": "prod"},
	}); err != nil {
		t.Fatalf("CreateSession #2: %v", err)
	}

	all, err := client.ListSessions(ctx, ListSessionsOptions{})
	if err != nil {
		t.Fatalf("ListSessions all: %v", err)
	}
	if all.Total != 2 {
		t.Fatalf("expected 2 sessions, got %d", all.Total)
	}

	filtered, err := client.ListSessions(ctx, ListSessionsOptions{
		Metadata: map[string]string{"customer": "acme"},
	})
	if err != nil {
		t.Fatalf("ListSessions filtered: %v", err)
	}
	if filtered.Total != 1 {
		t.Fatalf("expected 1 filtered session, got %d", filtered.Total)
	}
	if filtered.Sessions[0].Metadata["customer"] != "acme" {
		t.Fatalf("unexpected metadata: %#v", filtered.Sessions[0].Metadata)
	}
	if filtered.Sessions[0].Status != "active" {
		t.Fatalf("unexpected status: %q", filtered.Sessions[0].Status)
	}

	none, err := client.ListSessions(ctx, ListSessionsOptions{
		Metadata: map[string]string{"customer": "acme", "env": "staging"},
	})
	if err != nil {
		t.Fatalf("ListSessions none: %v", err)
	}
	if none.Total != 0 {
		t.Fatalf("expected 0 sessions, got %d", none.Total)
	}
}

func TestGetSessionEvents_ReplaysFrames(t *testing.T) {
	server := newMockServer()
	defer server.close()
	client := NewClient(Options{
		APIKey:        "k",
		WorkspaceSlug: "demo",
		BaseURL:       server.baseURL(),
	})
	ctx := context.Background()

	session, err := client.CreateSession(ctx, SessionSpec{SystemPrompt: "x"})
	if err != nil {
		t.Fatalf("CreateSession: %v", err)
	}
	if _, err := session.Send(ctx, "one"); err != nil {
		t.Fatalf("Send #1: %v", err)
	}
	if _, err := session.Send(ctx, "three"); err != nil {
		t.Fatalf("Send #2: %v", err)
	}

	full, err := client.GetSessionEvents(ctx, session.ID, GetSessionEventsOptions{})
	if err != nil {
		t.Fatalf("GetSessionEvents full: %v", err)
	}
	want := []struct {
		seq    int
		evType string
		text   string
	}{
		{1, "user_message", "one"},
		{2, "assistant_message", "echo:one"},
		{3, "user_message", "three"},
		{4, "assistant_message", "echo:three"},
	}
	if len(full) != len(want) {
		t.Fatalf("expected %d events, got %d (%#v)", len(want), len(full), full)
	}
	for i, w := range want {
		if full[i].Seq != w.seq || full[i].Type != w.evType {
			t.Fatalf("event %d: got {seq:%d type:%q}, want {seq:%d type:%q}", i, full[i].Seq, full[i].Type, w.seq, w.evType)
		}
		if txt, _ := full[i].Data["text"].(string); txt != w.text {
			t.Fatalf("event %d text: got %q, want %q", i, txt, w.text)
		}
	}

	lastTwo, err := session.Events(ctx, GetSessionEventsOptions{LastMessages: 2})
	if err != nil {
		t.Fatalf("Events lastMessages: %v", err)
	}
	if len(lastTwo) != 2 || lastTwo[0].Seq != 3 || lastTwo[1].Seq != 4 {
		t.Fatalf("unexpected lastMessages slice: %#v", lastTwo)
	}
}

func TestSession_RunGuardsOnCreateAndPerMessage(t *testing.T) {
	server := newMockServer()
	defer server.close()
	client := NewClient(Options{
		APIKey:        "k",
		WorkspaceSlug: "demo",
		BaseURL:       server.baseURL(),
	})

	session, err := client.CreateSession(context.Background(), SessionSpec{
		SystemPrompt:  "be helpful",
		LoopDetection: LoopDetectionThresholds(2, 4),
		ToolBudgets:   ToolBudgets{"recall": {MaxCalls: 4}},
	})
	if err != nil {
		t.Fatalf("CreateSession: %v", err)
	}
	var createBody map[string]any
	if err := json.Unmarshal(server.lastSessionCreateBody, &createBody); err != nil {
		t.Fatalf("parse create body: %v", err)
	}
	ld, _ := createBody["loopDetection"].(map[string]any)
	if ld["consecutiveThreshold"].(float64) != 2 || ld["hardCutoffThreshold"].(float64) != 4 {
		t.Fatalf("session loopDetection not forwarded: %#v", createBody["loopDetection"])
	}
	tb, _ := createBody["toolBudgets"].(map[string]any)
	recall, _ := tb["recall"].(map[string]any)
	if recall["maxCalls"].(float64) != 4 {
		t.Fatalf("session toolBudgets not forwarded: %#v", createBody["toolBudgets"])
	}

	if _, err := session.Send(context.Background(), "hi",
		WithLoopDetection(LoopDetectionDisabled()),
		WithToolBudgets(ToolBudgets{}),
	); err != nil {
		t.Fatalf("Send: %v", err)
	}
	var msgBody map[string]any
	if err := json.Unmarshal(server.lastSessionMessageBody, &msgBody); err != nil {
		t.Fatalf("parse message body: %v", err)
	}
	if got, ok := msgBody["loopDetection"].(bool); !ok || got != false {
		t.Fatalf("expected loopDetection=false override, got %#v", msgBody["loopDetection"])
	}
	mtb, ok := msgBody["toolBudgets"].(map[string]any)
	if !ok || len(mtb) != 0 {
		t.Fatalf("expected empty toolBudgets override, got %#v", msgBody["toolBudgets"])
	}
}
